File: packages/python/src/axonpush/resources/organizations.py
"""Organizations resource — current org, membership, invitations, admin."""

from __future__ import annotations

from typing import TYPE_CHECKING, List

from axonpush._internal.api.api.organization import (
    organization_delete as _delete_op,
    organization_get as _get_op,
    organization_invitations_accept as _accept_invite_op,
    organization_invitations_cancel as _delete_invite_op,
    organization_invitations_create as _invite_op,
    organization_leave as _leave_op,
    organization_members_remove as _remove_member_op,
    organization_transfer_ownership as _transfer_op,
    organization_update as _update_op,
)
from axonpush._internal.api.api.user import (
    user_me_organizations as _list_op,
)
from axonpush._internal.api.models import (
    AcceptInvitationInputBody,
    AcceptInvitationOutputBody,
    CreateInvitationInputBody,
    CreateInvitationInputBodyRole,
    InvitationDTO,
    OkOutputBody,
    TransferOwnershipInputBody,
    UpdateOrganizationInputBody,
    UserOrgsOutputBody,
)
from axonpush._internal.api.types import UNSET
from axonpush.models import Organization, UserOrg

if TYPE_CHECKING:
    from axonpush.resources._base import AsyncClientProtocol, SyncClientProtocol


def _unwrap(result: UserOrgsOutputBody | None) -> List[UserOrg] | None:
    if result is None:
        return None
    return list(result.organizations or [])


def _build_update_dto(*, name: str | None, description: str | None) -> UpdateOrganizationInputBody:
    return UpdateOrganizationInputBody(
        name=name if name is not None else UNSET,
        description=description if description is not None else UNSET,
    )


def _build_invite_dto(
    *, email: str, role: CreateInvitationInputBodyRole | str
) -> CreateInvitationInputBody:
    desired = (
        role
        if isinstance(role, CreateInvitationInputBodyRole)
        else CreateInvitationInputBodyRole(role)
    )
    return CreateInvitationInputBody(invited_email=email, role=desired)


class Organizations:
    """Synchronous organization management."""

    def __init__(self, client: SyncClientProtocol) -> None:
        self._client = client

    def list(self) -> List[UserOrg] | None:
        """List organizations the caller is a member of (envelope unwrapped).

        Backed by ``GET /users/me/organizations``.
        """
        return self._client._invoke(_list_op, _coerce=_unwrap)

    def get(self) -> Organization | None:
        """Fetch the caller's active organization. ``GET /organization``."""
        return self._client._invoke(_get_op)

    def update(
        self,
        *,
        name: str | None = None,
        description: str | None = None,
    ) -> Organization | None:
        """Edit the active organization's name / description. ``PATCH /organization``."""
        return self._client._invoke(
            _update_op, body=_build_update_dto(name=name, description=description)
        )

    def delete(self, org_id: str) -> OkOutputBody | None:
        """Delete an organization. ``DELETE /organizations/{orgId}``."""
        return self._client._invoke(_delete_op, org_id=org_id)

    def invite(
        self,
        org_id: str,
        email: str,
        *,
        role: CreateInvitationInputBodyRole | str = CreateInvitationInputBodyRole.USER,
    ) -> InvitationDTO | None:
        """Invite a user to an organization."""
        return self._client._invoke(
            _invite_op, org_id=org_id, body=_build_invite_dto(email=email, role=role)
        )

    def cancel_invitation(self, org_id: str, invitation_id: str) -> OkOutputBody | None:
        """Cancel a pending invitation."""
        return self._client._invoke(
            _delete_invite_op, org_id=org_id, invitation_id=invitation_id
        )

    def remove_member(self, org_id: str, user_id: str) -> OkOutputBody | None:
        """Remove a member from an organization."""
        return self._client._invoke(_remove_member_op, org_id=org_id, user_id=user_id)

    def transfer_ownership(self, org_id: str, target_user_id: str) -> OkOutputBody | None:
        """Transfer organization ownership to another member."""
        return self._client._invoke(
            _transfer_op,
            org_id=org_id,
            body=TransferOwnershipInputBody(user_id=target_user_id),
        )

    def leave(self, org_id: str) -> OkOutputBody | None:
        """Leave an organization. ``POST /organizations/{orgId}/leave``"""
        return self._client._invoke(_leave_op, org_id=org_id)

    def accept_invitation(
        self, body: AcceptInvitationInputBody
    ) -> AcceptInvitationOutputBody | None:
        """Accept a pending invitation. ``POST /invitations/accept``"""
        return self._client._invoke(_accept_invite_op, body=body)


class AsyncOrganizations:
    """Async sibling of :class:`Organizations`."""

    def __init__(self, client: AsyncClientProtocol) -> None:
        self._client = client

    async def list(self) -> List[UserOrg] | None:
        """See :meth:`Organizations.list`."""
        return await self._client._invoke(_list_op, _coerce=_unwrap)

    async def get(self) -> Organization | None:
        """See :meth:`Organizations.get`."""
        return await self._client._invoke(_get_op)

    async def update(
        self,
        *,
        name: str | None = None,
        description: str | None = None,
    ) -> Organization | None:
        """See :meth:`Organizations.update`."""
        return await self._client._invoke(
            _update_op, body=_build_update_dto(name=name, description=description)
        )

    async def delete(self, org_id: str) -> OkOutputBody | None:
        """See :meth:`Organizations.delete`."""
        return await self._client._invoke(_delete_op, org_id=org_id)

    async def invite(
        self,
        org_id: str,
        email: str,
        *,
        role: CreateInvitationInputBodyRole | str = CreateInvitationInputBodyRole.USER,
    ) -> InvitationDTO | None:
        """See :meth:`Organizations.invite`."""
        return await self._client._invoke(
            _invite_op, org_id=org_id, body=_build_invite_dto(email=email, role=role)
        )

    async def cancel_invitation(self, org_id: str, invitation_id: str) -> OkOutputBody | None:
        """See :meth:`Organizations.cancel_invitation`."""
        return await self._client._invoke(
            _delete_invite_op, org_id=org_id, invitation_id=invitation_id
        )

    async def remove_member(self, org_id: str, user_id: str) -> OkOutputBody | None:
        """See :meth:`Organizations.remove_member`."""
        return await self._client._invoke(_remove_member_op, org_id=org_id, user_id=user_id)

    async def transfer_ownership(self, org_id: str, target_user_id: str) -> OkOutputBody | None:
        """See :meth:`Organizations.transfer_ownership`."""
        return await self._client._invoke(
            _transfer_op,
            org_id=org_id,
            body=TransferOwnershipInputBody(user_id=target_user_id),
        )

    async def leave(self, org_id: str) -> OkOutputBody | None:
        """See :meth:`Organizations.leave`."""
        return await self._client._invoke(_leave_op, org_id=org_id)

    async def accept_invitation(
        self, body: AcceptInvitationInputBody
    ) -> AcceptInvitationOutputBody | None:
        """See :meth:`Organizations.accept_invitation`."""
        return await self._client._invoke(_accept_invite_op, body=body)

File: packages/typescript/src/resources/organizations.ts
import {
  organizationDelete,
  organizationGet,
  organizationInvitationsCancel,
  organizationInvitationsCreate,
  organizationMembersRemove,
  organizationTransferOwnership,
  organizationUpdate,
  userMeOrganizations,
} from "../_internal/api/sdk.gen.js";
import type {
  CreateInvitationInputBodyWritable,
  InvitationDto,
  OkOutputBody,
  UpdateOrganizationInputBodyWritable,
} from "../_internal/api/types.gen.js";
import type { Organization, UserOrg } from "../models.js";
import type { ResourceClient } from "./_client.js";

/** Mutable organization fields accepted by {@link OrganizationsResource.update}. */
export type OrganizationUpdateFields = UpdateOrganizationInputBodyWritable;

/** Invitation roles accepted by {@link OrganizationsResource.invite}. */
export type InvitationRole = CreateInvitationInputBodyWritable["role"];

/**
 * Read and manage organizations, invitations, and ownership transfers.
 *
 * Organization creation is owned by the auth service and is no longer part
 * of this SDK. `get`/`update` operate on the caller's active organization.
 */
export class OrganizationsResource {
  constructor(private readonly client: ResourceClient) {}

  /**
   * The active organization. `GET /organization`
   *
   * @returns The organization, or `null` on fail-open error.
   */
  async get(): Promise<Organization | null> {
    return this.client.invoke(organizationGet, {});
  }

  /**
   * List organizations the caller is a member of. `GET /users/me/organizations`
   *
   * @returns Membership records (org + role), or `null` on fail-open error.
   */
  async list(): Promise<UserOrg[] | null> {
    const res = await this.client.invoke(userMeOrganizations, {});
    return res?.organizations ?? null;
  }

  /**
   * Update the active organization. `PATCH /organization`
   *
   * @param fields - Patch object.
   * @returns The updated organization, or `null` on fail-open error.
   */
  async update(fields: OrganizationUpdateFields): Promise<Organization | null> {
    return this.client.invoke(organizationUpdate, { body: fields });
  }

  /**
   * Delete an organization. `DELETE /organizations/{orgId}`
   *
   * @param orgId - Organization id.
   * @returns Server ack, or `null` on fail-open error.
   */
  async delete(orgId: string): Promise<OkOutputBody | null> {
    return this.client.invoke(organizationDelete, { path: { orgId } });
  }

  /**
   * Invite a user to an organization. `POST /organizations/{orgId}/invitations`
   *
   * @param orgId - Organization id.
   * @param email - Invitee email.
   * @param role - Invitation role; defaults to `"user"`.
   * @returns The created invitation, or `null` on fail-open error.
   */
  async invite(
    orgId: string,
    email: string,
    role: InvitationRole = "user",
  ): Promise<InvitationDto | null> {
    return this.client.invoke(organizationInvitationsCreate, {
      path: { orgId },
      body: { invitedEmail: email, role },
    });
  }

  /**
   * Cancel a pending invitation.
   * `DELETE /organizations/{orgId}/invitations/{invitationId}`
   *
   * @param orgId - Organization id.
   * @param invitationId - Invitation id to cancel.
   * @returns Server ack, or `null` on fail-open error.
   */
  async cancelInvitation(orgId: string, invitationId: string): Promise<OkOutputBody | null> {
    return this.client.invoke(organizationInvitationsCancel, {
      path: { orgId, invitationId },
    });
  }

  /**
   * Remove a member from an organization.
   * `DELETE /organizations/{orgId}/members/{userId}`
   *
   * @param orgId - Organization id.
   * @param userId - User id to remove.
   * @returns Server ack, or `null` on fail-open error.
   */
  async removeMember(orgId: string, userId: string): Promise<OkOutputBody | null> {
    return this.client.invoke(organizationMembersRemove, {
      path: { orgId, userId },
    });
  }

  /**
   * Transfer ownership of an organization to another member.
   * `POST /organizations/{orgId}/transfer-ownership`
   *
   * @param orgId - Organization id.
   * @param targetUserId - User id to promote.
   * @returns Server ack, or `null` on fail-open error.
   */
  async transferOwnership(orgId: string, targetUserId: string): Promise<OkOutputBody | null> {
    return this.client.invoke(organizationTransferOwnership, {
      path: { orgId },
      body: { userId: targetUserId },
    });
  }
}
